File: haskell/sublist/Sublist.hs
module Sublist(Sublist(..), sublist) where

import Data.List(tails)

data Sublist = Equal | Sublist | Superlist | Unequal deriving(Eq, Show)

sublist :: (Eq a) => [a] -> [a] -> Sublist
sublist [] [] = Equal
sublist [] _  = Sublist
sublist xs ys
    | lxs  > lys = invert (sublist ys xs)
    | lxs == lys && prefix xs ys = Equal
    | any (prefix xs) (tails ys) = Sublist
    | otherwise = Unequal
    where
        lxs = length xs
        lys = length ys
        invert s
            | s == Sublist = Superlist
            | s == Superlist = Sublist
            | otherwise = s
        prefix [] _ = True
        prefix _ [] = False
        prefix (x:xss) (y:yss) = x == y && prefix xss yss

File: python/sublist/sublist.py
from enum import IntEnum


class SublistType(IntEnum):
    SUBLIST, SUPERLIST, EQUAL, UNEQUAL = range(4)

    def inverse(self):
        if self == SUBLIST:
            return SUPERLIST
        if self == SUPERLIST:
            return SUBLIST
        return self


SUBLIST, SUPERLIST = SublistType.SUBLIST, SublistType.SUPERLIST
EQUAL, UNEQUAL = SublistType.EQUAL, SublistType.UNEQUAL


def check_lists(A, B):
    length_a, length_b = len(A), len(B)

    if length_a > length_b:
        return check_lists(B, A).inverse()

    if A == B:
        return EQUAL

    for k in range(length_b - length_a + 1):
        if A == B[k: length_a + k]:
            return SUBLIST

    return UNEQUAL
